kubernetes/provider_test.go

```diff
@@ -4,44 +4,62 @@ import (
 	"context"
 	"errors"
 	"fmt"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
 	"os"
 	"strings"
 	"testing"
 
 	gversion "github.com/hashicorp/go-version"
-	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
-	"github.com/hashicorp/terraform-plugin-sdk/terraform"
-	"github.com/hashicorp/terraform-provider-google/google"
-	"github.com/terraform-providers/terraform-provider-aws/aws"
-	"github.com/terraform-providers/terraform-provider-azurerm/azurerm"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/terraform"
 	api "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
-var testAccProviders map[string]terraform.ResourceProvider
+var testAccProviders map[string]*schema.Provider
 var testAccProvider *schema.Provider
+var testAccExternalProviders map[string]resource.ExternalProvider
+var testAccProviderFactories = map[string]func() (*schema.Provider, error){
+	"kubernetes": func() (*schema.Provider, error) {
+		return Provider(), nil
+	},
+}
 
 func init() {
-	testAccProvider = Provider().(*schema.Provider)
-	testAccProviders = map[string]terraform.ResourceProvider{
+	testAccProvider = Provider()
+	testAccProviders = map[string]*schema.Provider{
 		"kubernetes": testAccProvider,
-		"google":     google.Provider(),
-		"aws":        aws.Provider(),
-		"azurerm":    azurerm.Provider(),
+	}
+	testAccProviderFactories = map[string]func() (*schema.Provider, error){
+		"kubernetes": func() (*schema.Provider, error) {
+			return Provider(), nil
+		},
+	}
+	testAccExternalProviders = map[string]resource.ExternalProvider{
+		"kubernetes-local": {
```

Comment:
yeah, I think I can do a better job here now that I have gone through this process. This was the first solution I came up with, so it's a little rough around the edges, but I bet I can improve it. I'll give it a try and let you know how it goes.

Reply:
darn, I was excited to tidy this up. But I don't think it's possible. Here's the test I ran to cause a ForceNew:

First, I made this change without modifying anything else. This ensures that we can catch a ForceNew with the code as it is right now:

```
kubernetes/resource_kubernetes_pod.go
+       podSpecFields["dns_policy"].Default = "Default"
+       podSpecFields["dns_policy"].ForceNew = true

 TESTARGS="-run 'TestAccKubernetesPod_regression'" make testacc
```
This was the result, which is good! It caught the ForceNew:

```
=== RUN   TestAccKubernetesPod_regression
    resource_kubernetes_pod_test.go:968: Step 2/2 error: Check failed: 2 errors occurred:
                * Check 11/34 error: kubernetes_pod.test: Attribute 'spec.0.dns_policy' expected "ClusterFirst", got "Default"
                * Check 34/34 error: Expecting pod UIDs to be the same: expected 17675004-ae7d-4d07-8713-bf530ed49a6b got bcdea13f-0c9c-4568-b7bd-1a2ab166c719
--- FAIL: TestAccKubernetesPod_regression (61.76s)
FAIL
```

For the second test, I tidied up the code by removing the "kubernetes-local" provider. It was encouraging to see that it could run without that. I just passed "kubernetes" as the provider name instead:

```
kubernetes/resource_kubernetes_pod_test.go
@@ -968,6 +968,7 @@ func TestAccKubernetesPod_regression(t *testing.T) {                                                                                
        resource.Test(t, resource.TestCase{                                                                                                             
                PreCheck:          func() { testAccPreCheck(t) },                                                                                       
                ExternalProviders: testAccExternalProviders,
+               ProviderFactories: testAccProviderFactories,

-                               Config: requiredProviders() + testAccKubernetesPod_regression("kubernetes-local", name, imageName),
+                               Config: requiredProviders() + testAccKubernetesPod_regression("kubernetes", name, imageName),
```

However, the results were not as expected. The test did not catch the ForceNew. I'm not sure why. The full diff is [here](https://gist.githubusercontent.com/dak1n1/2e82808221842a56acd5e66b3e6fc593/raw/c0a47b868abf6d5181afbbaa9a88536f8bcd6d2a/gistfile1.txt), in case you want to look. 

```
=== RUN   TestAccKubernetesPod_regression
    resource_kubernetes_pod_test.go:968: Step 1/2 error: Check failed: 1 error occurred:
                * Check 11/33 error: kubernetes_pod.test: Attribute 'spec.0.dns_policy' expected "ClusterFirst", got "Default"
        
--- FAIL: TestAccKubernetesPod_regression (23.22s)
```

The reason I initially compiled the provider as an ExternalProvider is because I wasn't able to get the "terraform init" behavior I needed without it. The terraform init is needed in between TestSteps to use the newer version of the provider and replace the released version (simulating an upgrade). This is what I'll open an issue for upstream. Because I agree that my solution is really more of a work-around, and it would be nicer to have the SDK support this use case.

Reply:
Awesome, thanks for investigating. This is something I'd like to see in the SDK, so this work is good evidence for a case for it.